src/machine/usb_nrf52840.go

```diff
@@ -0,0 +1,541 @@
+			nrf.USBD.TASKS_EP0STATUS.Set(1)
+		}
+		return
+	}
+
+	// do not know how to handle this message, so return zero
+	nrf.USBD.TASKS_EP0STATUS.Set(1)
+	return
+}
+
+// sendConfiguration creates and sends the configuration packet to the host.
+func sendConfiguration(setup usbSetup) {
+	if setup.wLength == 9 {
+		sz := uint16(configDescriptorSize + cdcSize)
+		config := NewConfigDescriptor(sz, 2)
+
+		sendUSBPacket(0, config.Bytes())
+	} else {
+		iad := NewIADDescriptor(0, 2, usb_CDC_COMMUNICATION_INTERFACE_CLASS, usb_CDC_ABSTRACT_CONTROL_MODEL, 0)
+
+		cif := NewInterfaceDescriptor(usb_CDC_ACM_INTERFACE, 1, usb_CDC_COMMUNICATION_INTERFACE_CLASS, usb_CDC_ABSTRACT_CONTROL_MODEL, 0)
+
+		header := NewCDCCSInterfaceDescriptor(usb_CDC_HEADER, usb_CDC_V1_10&0xFF, (usb_CDC_V1_10>>8)&0x0FF)
+
+		controlManagement := NewACMFunctionalDescriptor(usb_CDC_ABSTRACT_CONTROL_MANAGEMENT, 6)
+
+		functionalDescriptor := NewCDCCSInterfaceDescriptor(usb_CDC_UNION, usb_CDC_ACM_INTERFACE, usb_CDC_DATA_INTERFACE)
+
+		callManagement := NewCMFunctionalDescriptor(usb_CDC_CALL_MANAGEMENT, 1, 1)
+
+		cifin := NewEndpointDescriptor((usb_CDC_ENDPOINT_ACM | usbEndpointIn), usb_ENDPOINT_TYPE_INTERRUPT, 0x10, 0x10)
+
+		dif := NewInterfaceDescriptor(usb_CDC_DATA_INTERFACE, 2, usb_CDC_DATA_INTERFACE_CLASS, 0, 0)
+
+		out := NewEndpointDescriptor((usb_CDC_ENDPOINT_OUT | usbEndpointOut), usb_ENDPOINT_TYPE_BULK, usbEndpointPacketSize, 0)
+
+		in := NewEndpointDescriptor((usb_CDC_ENDPOINT_IN | usbEndpointIn), usb_ENDPOINT_TYPE_BULK, usbEndpointPacketSize, 0)
+
+		cdc := NewCDCDescriptor(iad,
+			cif,
+			header,
+			controlManagement,
+			functionalDescriptor,
+			callManagement,
+			cifin,
+			dif,
+			out,
+			in)
+
+		sz := uint16(configDescriptorSize + cdcSize)
+		config := NewConfigDescriptor(sz, 2)
+
+		buf := make([]byte, 0)
+		buf = append(buf, config.Bytes()...)
+		buf = append(buf, cdc.Bytes()...)
+		sendUSBPacket(0, buf)
+	}
+}
+
+func handleEndpoint(ep uint32) {
+	// get data
+	count := int(nrf.USBD.SIZE.EPOUT[ep].Get())
+	// move to ring buffer
+	for i := 0; i < count; i++ {
+		UART0.Receive(byte((udd_ep_out_cache_buffer[ep][i] & 0xFF)))
+	}
+
+	// set ready for next data
+	nrf.USBD.SIZE.EPOUT[ep].Set(0)
+	nrf.USBD.TASKS_STARTEPOUT[ep].Set(1)
+}
+
+func sendViaEPIn(ep uint32, ptr *byte, count int) {
+	nrf.USBD.EPIN[ep].PTR.Set(
+		uint32(uintptr(unsafe.Pointer(ptr))),
+	)
+	nrf.USBD.EPIN[ep].MAXCNT.Set(uint32(count))
+	nrf.USBD.EVENTS_ENDEPIN[ep].Set(0)
+	nrf.USBD.TASKS_STARTEPIN[ep].Set(1)
+}
+
+func enableEPOut(ep uint32) {
+	epouten = epouten | (nrf.USBD_EPOUTEN_OUT0 << ep)
+	nrf.USBD.EPOUTEN.Set(epouten)
+}
+
+func enableEPIn(ep uint32) {
+	epinen = epinen | (nrf.USBD_EPINEN_IN0 << ep)
+	nrf.USBD.EPINEN.Set(epinen)
+}
+
+func strToUTF16Descriptor(in string) []byte {
+	size := (len(in)<<1) + 2
+	out := make([]byte, size)
+	out[0] = byte(size)
+	out[1] = 0x03
+	// TODO: do a "real" conversion that supports more than just ascii chars
+	for i, char := range in {
```

Comment:
Note that you're already decoding Unicode code points here: `char` is actually a single Unicode code point (and `i` is the index into the string, which will increase by more than 1 when decoding non-ASCII characters).

Reply:
I just copied what was done [here](https://github.com/tinygo-org/tinygo/blob/master/src/machine/machine_atsamd51.go#L1877-L1885) except I'm iterating runes instead of bytes. Neither this implementation or the other will work for anything except ascii characters. I was worried about pulling in the `unicode` package to convert, but it looks like I can do that fairly easily.

Reply:
Oh, it was more of a hint. And yeah, if possible it would be best to avoid the unicode package (it is really big and currently slow to compile).